src/coordination/coordinator_instance.cpp

```diff
@@ -214,78 +223,148 @@ auto CoordinatorInstance::ShowInstances() const -> std::vector<InstanceStatus> {
+                                                   // not ready still as leader
     }
-  } else {
-    auto const stringify_inst_status =
-        [raft_state_ptr = raft_state_.get()](ReplicationInstanceState const &instance) -> std::string {
-      if (raft_state_ptr->IsCurrentMain(instance.config.instance_name)) {
-        return "main";
-      }
-      if (raft_state_ptr->HasMainState(instance.config.instance_name)) {
-        return "unknown";
+
+    if (leader_id == -1) {
+      spdlog::trace("No leader found, returning report as follower {}", std::this_thread::get_id());
+      return {ShowInstancesState::FOLLOWER, GetAllInstancesStatusAsFollower()};
+    }
+    CoordinatorInstanceConnector *follower{nullptr};
+    // TODO(antoniofilipovic) move this part to separate function
+    {
+      auto connectors = coordinator_connectors_.Lock();
+      if (!connectors->contains(leader_id)) {
+        auto all_coordinators = raft_state_->GetCoordinatorInstances();
+
+        auto coord = std::ranges::find_if(all_coordinators,
+                                          [leader_id](auto const &coord) { return coord.coordinator_id == leader_id; });
+
+        MG_ASSERT(coord != all_coordinators.end());
+        connectors->emplace(
+            leader_id, CoordinatorInstanceConnector{CoordinatorInstanceManagementServerConfig{coord->management_server},
+                                                    static_cast<int>(leader_id)});
       }
-      return "replica";
-    };
+      follower = &connectors->at(leader_id);
+    }
 
-    auto process_repl_instance_as_follower =
-        [&stringify_inst_status](ReplicationInstanceState const &instance) -> InstanceStatus {
-      return {.instance_name = instance.config.instance_name,
-              .management_server = instance.config.ManagementSocketAddress(),  // show non-resolved IP
-              .bolt_server = instance.config.BoltSocketAddress(),              // show non-resolved IP
-              .cluster_role = stringify_inst_status(instance),
-              .health = "unknown"};
-    };
+    MG_ASSERT(follower != nullptr, "Follower not found");
+
+    auto maybe_res = follower->SendShowInstances();
 
-    std::ranges::transform(raft_state_->GetReplicationInstances(), std::back_inserter(instances_status),
-                           process_repl_instance_as_follower);
+    if (!maybe_res.has_value()) {
+      spdlog::trace("Couldn't get instances from leader, returning report as follower {}", std::this_thread::get_id());
+      return {ShowInstancesState::FOLLOWER, GetAllInstancesStatusAsFollower()};
+    }
+    spdlog::trace("Got instances from leader, returning report as leader.");
+    return {ShowInstancesState::LEADER, std::move(maybe_res.value())};
   }
-  return instances_status;
+  spdlog::trace("Processing show instances as leader {}", std::this_thread::get_id());
+  auto instances_status = GetCoordinatorsInstanceStatus();
```

Comment:
In my view this method ShowInstances can be made simpler:

```
if (my_coordinator_id == current_leader_id) {
 return ShowInstancesFromLeader();
}
if (current_leader_exists) {
 auto optional_res =  GetShowInstancesFromLeader(); // this sends rpc request inside. It searches over all connections
// and finds RPC connection by searching leader_id. Having map of connections is probably an overkill sincer
// there are just several connection (=num of coordinators). So you can use simplest possible = vector.
// It could happen that when instance receives RPC request it changed role from leader to follower. That is fine, // we can either:
// 1. Try to send request to new leader (circular, I like it, shouldn't be a problem)
// 2. Instance can return info as if it were follower -> also completely fine, we just document it
if (optional_res) {
return *optional_res;
}
}
return ShowInstancesAsFollower(); // RPC failed or we cannot find current leader
```

Reply:
Also what we need to take care of. We should just find who is the current leader once, at the beginning of show instances and then propagate this through methods so we get a feeling of immutability. As it is written now, we're finding leader multiple times which could be a problem. Again, simplest possible is good.

Reply:
The reasoning not for option 1. is the following:
What if the coordinator which is just becoming the leader is a follower and gets SHOW INSTANCES
 - it will try to contact that leader
 - the old leader will see that it is not the leader and it will try to contact the new leader which is that instance
 - that instance is in the process of sending the RPC request so it can't send a new one to answer to that question so you are in the state of deadlock

Option 2. is better, but I would try to answer as a leader

Reply:
okay that's fine, return as follower